{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# import the nessessary packages\n",
    "import pickle\n",
    "import os\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import polars as pl\n",
    "import seaborn as sns\n",
    "import plotly.graph_objects as go\n",
    "from models.ecoli.analysis import variantAnalysisPlot\n",
    "from wholecell.analysis.analysis_tools import (exportFigure,\n",
    "\tread_bulk_molecule_counts, read_stacked_bulk_molecules, read_stacked_columns)\n",
    "from wholecell.io.tablereader import TableReader"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# work with the filterd data from saved_data_ng_internal_shift1: \n",
    "os.chdir(os.path.expanduser('~/wcEcoli/out/sherlock_data/saved_data_ng_internal_shift1/filtered_data/'))\n",
    "\n",
    "# work specifically with the log data for the second sherlock run \n",
    "name_front = 'Filtered_AvgProteinCounts_Variant_'\n",
    "name_back = '_startGen_14.csv'\n",
    "\n",
    "# access the unfiltered protein counts too so that the total PC divisor can be calculated: \n",
    "unfiltered_pth = '~/wcEcoli/out/sherlock_data/saved_data_ng_internal_shift1/unfiltered_data/'\n",
    "name_front_unfiltered = 'AvgProteinCounts_Variant_'\n",
    "\n",
    "# load the varaint info:\n",
    "variant_info = pd.read_excel('~/wcEcoli/out/sherlock_data/saved_data_ng_internal_shift1/sherlock_sim1_info.xlsx')\n",
    "variant_info = pl.DataFrame(variant_info)"
   ],
   "id": "22d069dfaad6be70",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# define a function that calculates the total protein count for each variant:\n",
    "def get_total_protein_counts(var_num):\n",
    "    # calculate the sum of all the protein counts for both the control and experimental variant (unfiltered):\n",
    "    unfiltered_df = pd.read_csv(unfiltered_pth + name_front_unfiltered + str(var_num) + name_back)\n",
    "    pl_unfiltered_df = pl.DataFrame(unfiltered_df)\n",
    "    pl_unfiltered_df = pl_unfiltered_df.rename({pl_unfiltered_df.columns[1]: \"Control_Variant\", pl_unfiltered_df.columns[2]: \"Experimental_Variant\"}) \n",
    "    df_sum = pl_unfiltered_df.sum()\n",
    "    \n",
    "    # get the log10 of the total protein counts for the control and experimental variant:\n",
    "    df_log10_Csum = df_sum[\"Control_Variant\"].log10()\n",
    "    df_log10_Esum = df_sum[\"Experimental_Variant\"].log10()\n",
    "    \n",
    "    return df_sum[\"Control_Variant\"].item(), df_sum[\"Experimental_Variant\"].item(), df_log10_Csum.item(), df_log10_Esum.item()\n",
    "\n",
    "# add a column to the variant info that contains the total protein counts for each variant:\n",
    "total_PCs = [get_total_protein_counts(1)[0]]\n",
    "for i in range(1, 21): \n",
    "    total_PCs.append(get_total_protein_counts(i)[1])\n",
    "variant_info = variant_info.with_columns(pl.Series(\"total protein counts\", total_PCs))\n",
    "variant_info = variant_info.with_columns(pl.col(\"total protein counts\").log10().alias(\"$log_{10}(total protein counts)\"))\n",
    "variant_info"
   ],
   "id": "a274a3cd5bfd2056",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Unfiltered Heatmpas",
   "id": "a3f6c577f3c3863c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the variant numbers: \n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"variant\"])\n",
    "df = df.to_pandas()\n",
    "\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values='variant')\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".0f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "# reverse the order of the y-axis:\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"Variant Number\")\n"
   ],
   "id": "a6653f1ff2fd63f2",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the total analysis sims used for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"analysis sims\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values='analysis sims')\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".0f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"Analysis Sims for Each Variant\")"
   ],
   "id": "2d85969739c55ff",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the total protein counts for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", 'total protein counts'])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values='total protein counts')\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".1f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"Total Protein Counts for Each Variant\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show()"
   ],
   "id": "1973f0e09706cf73",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"$log_{10}(total protein counts)\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"$log_{10}(total protein counts)\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".2f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"$log_{10}$(total protein counts) for Each Variant\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show()"
   ],
   "id": "71dfd730bd696a7",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# find the max and min normalized protein count values for each variant: \n",
    "def get_max_min_log10normalized_protein_counts(var_num):\n",
    "    \n",
    "    # calculate the sum of all the protein counts for both the control and experimental variant (unfiltered):\n",
    "    unfiltered_df = pd.read_csv(unfiltered_pth + name_front_unfiltered + str(var_num) + name_back)\n",
    "    pl_unfiltered_df = pl.DataFrame(unfiltered_df)\n",
    "    pl_unfiltered_df = pl_unfiltered_df.rename({pl_unfiltered_df.columns[1]: \"Control_Variant\", pl_unfiltered_df.columns[2]: \"Experimental_Variant\"}) \n",
    "    df_sum = pl_unfiltered_df.sum()\n",
    "  \n",
    "    # normalize the control data and take the log10 of the data (note: log10(protein_count_value) - log10(sum_over_all_protein_counts) = log10(protein_count_value/sum_over_all_protein_counts)):\n",
    "    df_log10_C = pl_unfiltered_df.with_columns((pl.col(\"Control_Variant\")+1).log10().alias(\"log10_Control_Variant\"))\n",
    "    df_log10_Csum = df_sum[\"Control_Variant\"].log10()\n",
    "    df_log10_C = df_log10_C.with_columns(pl.col(\"log10_Control_Variant\") - df_log10_Csum)\n",
    "    \n",
    "    # normalize the experimental data and take the log10 of the data :\n",
    "    df_log10_E = df_log10_C.with_columns((pl.col(\"Experimental_Variant\")+1).log10().alias(\"log10_Experimental_Variant\"))\n",
    "    df_log10_Esum = df_sum[\"Experimental_Variant\"].log10()\n",
    "    df_log10_E = df_log10_E.with_columns(pl.col(\"log10_Experimental_Variant\") - df_log10_Esum)\n",
    "    \n",
    "    # get the max and min values for the normalized protein counts:\n",
    "    max_C = df_log10_C[\"log10_Control_Variant\"].max()\n",
    "    min_C = df_log10_C[\"log10_Control_Variant\"].min()\n",
    "    max_E = df_log10_E[\"log10_Experimental_Variant\"].max()\n",
    "    min_E = df_log10_E[\"log10_Experimental_Variant\"].min()\n",
    "    \n",
    "    return max_C, min_C, max_E, min_E\n",
    "\n",
    "def get_max_min_normalized_protein_counts(var_num):\n",
    "    \n",
    "    # calculate the sum of all the protein counts for both the control and experimental variant (unfiltered):\n",
    "    unfiltered_df = pd.read_csv(unfiltered_pth + name_front_unfiltered + str(var_num) + name_back)\n",
    "    pl_unfiltered_df = pl.DataFrame(unfiltered_df)\n",
    "    pl_unfiltered_df = pl_unfiltered_df.rename({pl_unfiltered_df.columns[1]: \"Control_Variant\", pl_unfiltered_df.columns[2]: \"Experimental_Variant\"}) \n",
    "    df_sum = pl_unfiltered_df.sum()\n",
    "  \n",
    "    # normalize the control data and take the log10 of the data (note: log10(protein_count_value) - log10(sum_over_all_protein_counts) = log10(protein_count_value/sum_over_all_protein_counts)):\n",
    "    df_C = pl_unfiltered_df.with_columns((pl.col(\"Control_Variant\") / df_sum[\"Control_Variant\"]).alias(\"normalized_Control_Variant\"))\n",
    "    \n",
    "    # normalize the experimental data and take the log10 of the data :\n",
    "    df_CE = df_C.with_columns((pl.col(\"Experimental_Variant\") / df_sum[\"Experimental_Variant\"]).alias(\"normalized_Experimental_Variant\"))\n",
    "    \n",
    "    \n",
    "    # get the max and min values for the normalized protein counts:\n",
    "    max_C = df_CE[\"normalized_Control_Variant\"].max()\n",
    "    min_C = df_CE[\"normalized_Control_Variant\"].min()\n",
    "    max_E = df_CE[\"normalized_Experimental_Variant\"].max()\n",
    "    min_E = df_CE[\"normalized_Experimental_Variant\"].min()\n",
    "    \n",
    "    return max_C, min_C, max_E, min_E\n"
   ],
   "id": "da785424cad1d9c0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# heat map of the max and min values for the log10 normalized protein counts for each variant:\n",
    "max_nums = [get_max_min_normalized_protein_counts(1)[0]]\n",
    "min_nums = [get_max_min_normalized_protein_counts(1)[1]]\n",
    "log_max_nums = [get_max_min_log10normalized_protein_counts(1)[0]]\n",
    "log_min_nums = [get_max_min_log10normalized_protein_counts(1)[1]]\n",
    "\n",
    "for i in range(1, 21):\n",
    "    get_max_min_normalized_protein_counts(i)\n",
    "    max_nums.append(get_max_min_normalized_protein_counts(i)[2])\n",
    "    min_nums.append(get_max_min_normalized_protein_counts(i)[3])\n",
    "    log_max_nums.append(get_max_min_log10normalized_protein_counts(i)[2])\n",
    "    log_min_nums.append(get_max_min_log10normalized_protein_counts(i)[3])\n",
    "    \n",
    "# add the data to the variant info:\n",
    "# add a column to the variant info that contains the total protein counts for each variant:\n",
    "\n",
    "variant_info = variant_info.with_columns(pl.Series(\"max normalized PC\", max_nums))\n",
    "variant_info = variant_info.with_columns(pl.Series(\"min normalized PC\", min_nums))\n",
    "variant_info = variant_info.with_columns(pl.Series(\"max log10 normalized PC\", log_max_nums))\n",
    "variant_info = variant_info.with_columns(pl.Series(\"min log10 normalized PC\", log_min_nums))\n",
    "variant_info\n"
   ],
   "id": "76698b4171220c2",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"max normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"max normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".3f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"max normalized PC for Each Variant\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show()"
   ],
   "id": "a03133e274b18b43",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"min normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"min normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".8f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"min normalized PC for Each Variant\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show() # looks just as expected"
   ],
   "id": "7d6fef98ec031ba",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# log max plot\n",
    "\n",
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"max log10 normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"max log10 normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".3f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"max log10 normalized PC for Each Variant\")"
   ],
   "id": "681bb403d369e367",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant: \n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info.select([\"NG expression\", \"TE\", \"min log10 normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"min log10 normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".8f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')    \n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"min log10 normalized PC for Each Variant\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show() # is this as expeccted?"
   ],
   "id": "e734db14c0190b09",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Filtered data (PCs > 0)",
   "id": "cbc066355afa22b8"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# get the filtered mins and maxs: \n",
    "def get_max_min_filtered_log10normalized_protein_counts(var_num):\n",
    "    # calculate the sum of all the protein counts for both the control and experimental variant (unfiltered):\n",
    "    filtered_df = pd.read_csv(name_front + str(var_num) + name_back)\n",
    "    pl_df = pl.DataFrame(filtered_df)\n",
    "    pl_df = pl_df.rename({pl_df.columns[1]: \"Control_Variant\", pl_df.columns[2]: \"Experimental_Variant\"}) \n",
    "  \n",
    "    CV_sum, EV_sum, CV_log_sum, EV_log_sum = get_total_protein_counts(var_num)\n",
    "    \n",
    "    # normalize the control data and take the log10 of the data (note: log10(protein_count_value) - log10(sum_over_all_protein_counts) = log10(protein_count_value/sum_over_all_protein_counts)):\n",
    "    df_log10_C = pl_df.with_columns((pl.col(\"Control_Variant\")).log10().alias(\"filtered_log10_Control_Variant\"))\n",
    "    df_log10_C = df_log10_C.with_columns(pl.col(\"filtered_log10_Control_Variant\") - CV_log_sum)\n",
    "    \n",
    "    # normalize the experimental data and take the log10 of the data :\n",
    "    df_log10_E = df_log10_C.with_columns((pl.col(\"Experimental_Variant\")).log10().alias(\"filtered_log10_Experimental_Variant\"))\n",
    "    df_log10_E = df_log10_E.with_columns(pl.col(\"filtered_log10_Experimental_Variant\") - EV_log_sum)\n",
    "    \n",
    "    # get the max and min values for the normalized protein counts:\n",
    "    max_C = df_log10_C[\"filtered_log10_Control_Variant\"].max()\n",
    "    min_C = df_log10_C[\"filtered_log10_Control_Variant\"].min()\n",
    "    max_E = df_log10_E[\"filtered_log10_Experimental_Variant\"].max()\n",
    "    min_E = df_log10_E[\"filtered_log10_Experimental_Variant\"].min()\n",
    "    \n",
    "    return max_C, min_C, max_E, min_E\n",
    "\n",
    "def get_max_min_filtered_normalized_protein_counts(var_num):\n",
    "    \n",
    "    # calculate the sum of all the protein counts for both the control and experimental variant (unfiltered):\n",
    "    filtered_df = pd.read_csv(name_front + str(var_num) + name_back)\n",
    "    pl_df = pl.DataFrame(filtered_df)\n",
    "    pl_df = pl_df.rename({pl_df.columns[1]: \"Control_Variant\", pl_df.columns[2]: \"Experimental_Variant\"}) \n",
    "  \n",
    "    CV_sum, EV_sum, CV_log_sum, EV_log_sum = get_total_protein_counts(var_num)\n",
    "  \n",
    "    # normalize the control data:\n",
    "    df_C = pl_df.with_columns((pl.col(\"Control_Variant\") / CV_sum).alias(\"filtered_normalized_Control_Variant\"))\n",
    "    \n",
    "    # normalize the experimental data data :\n",
    "    df_CE = df_C.with_columns((pl.col(\"Experimental_Variant\") / EV_sum).alias(\"filtered_normalized_Experimental_Variant\"))\n",
    "    \n",
    "    # get the max and min values for the normalized protein counts:\n",
    "    max_C = df_CE[\"filtered_normalized_Control_Variant\"].max()\n",
    "    min_C = df_CE[\"filtered_normalized_Control_Variant\"].min()\n",
    "    max_E = df_CE[\"filtered_normalized_Experimental_Variant\"].max()\n",
    "    min_E = df_CE[\"filtered_normalized_Experimental_Variant\"].min()\n",
    "    \n",
    "    return max_C, min_C, max_E, min_E\n"
   ],
   "id": "da8282026deb20f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# heat map of the max and min values for the log10 normalized protein counts for each variant:\n",
    "max_nums = [get_max_min_filtered_normalized_protein_counts(1)[0]]\n",
    "min_nums = [get_max_min_filtered_normalized_protein_counts(1)[1]]\n",
    "log_max_nums = [get_max_min_filtered_log10normalized_protein_counts(1)[0]]\n",
    "log_min_nums = [get_max_min_filtered_log10normalized_protein_counts(1)[1]]\n",
    "\n",
    "for i in range(1, 21):\n",
    "    max_nums.append(get_max_min_filtered_normalized_protein_counts(i)[2])\n",
    "    min_nums.append(get_max_min_filtered_normalized_protein_counts(i)[3])\n",
    "    log_max_nums.append(get_max_min_filtered_log10normalized_protein_counts(i)[2])\n",
    "    log_min_nums.append(get_max_min_filtered_log10normalized_protein_counts(i)[3])\n",
    "    \n",
    "# add the data to the variant info:\n",
    "# add a column to the variant info that contains the total protein counts for each variant:\n",
    "\n",
    "variant_info_filtered = variant_info.with_columns(pl.Series(\"max filtered normalized PC\", max_nums))\n",
    "variant_info_filtered = variant_info_filtered.with_columns(pl.Series(\"min filtered normalized PC\", min_nums))\n",
    "variant_info_filtered = variant_info_filtered.with_columns(pl.Series(\"max filtered log10 normalized PC\", log_max_nums))\n",
    "variant_info_filtered = variant_info_filtered.with_columns(pl.Series(\"min filtered log10 normalized PC\", log_min_nums))\n",
    "variant_info_filtered\n"
   ],
   "id": "24892c5978dd04c8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info_filtered.select([\"NG expression\", \"TE\", \"max filtered normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"max filtered normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".3f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"max normalized PC for Each Variant (filtered > 0)\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show()"
   ],
   "id": "a7d5fb99e817baf8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info_filtered.select([\"NG expression\", \"TE\", \"min filtered normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"min filtered normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".8f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"min normalized PC for Each Variant (filtered > 0)\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show() # looks just as expected"
   ],
   "id": "43960e060c8fd638",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# log max plot\n",
    "\n",
    "# make a heatmap of the log(total protein counts) for each variant:\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info_filtered.select([\"NG expression\", \"TE\", \"max filtered log10 normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"max filtered log10 normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".3f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')\n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"max log10 normalized PC for Each Variant (filtered > 0)\")"
   ],
   "id": "1074974935574a55",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# make a heatmap of the log(total protein counts) for each variant: \n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "df = variant_info_filtered.select([\"NG expression\", \"TE\", \"min filtered log10 normalized PC\"])\n",
    "df = df.to_pandas()\n",
    "heatmap_data = df.pivot(index='TE', columns='NG expression', values=\"min filtered log10 normalized PC\")\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.heatmap(heatmap_data, annot=True, fmt=\".8f\", cmap=\"viridis\", cbar=True, linewidths=.5, linecolor='gray')    \n",
    "plt.gca().invert_yaxis()\n",
    "\n",
    "# change the x and y labels:\n",
    "plt.xlabel(\"New Gene Expression Index\")\n",
    "plt.ylabel(\"New Gene Translation Efficiency\")\n",
    "plt.title(\"min log10 normalized PC for Each Variant (filtered > 0)\")\n",
    "\n",
    "# Display the heatmap\n",
    "plt.show() # is this as expeccted?"
   ],
   "id": "df634e115eea87fa",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "e46102628177431b",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
