{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# CLClim0NE: Visualize the results of the CLClim0NE protein outputs compared to those in the original model",
   "id": "3d7428944e1fdff2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pickle\n",
    "import os\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "os.chdir(os.path.expanduser('~/wcEcoli/out/'))\n",
    "# noinspection PyUnresolvedReferences\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "from models.ecoli.analysis import cohortAnalysisPlot\n",
    "from wholecell.analysis.analysis_tools import (exportFigure,\n",
    "\tread_bulk_molecule_counts, read_stacked_bulk_molecules, read_stacked_columns)\n",
    "from wholecell.io.tablereader import TableReader\n",
    "from sklearn.metrics import r2_score"
   ],
   "id": "initial_id",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Load in the simulation data for the original model (CLNE) and the CLClim2NE model",
   "id": "5c46ecb7fe0b1932"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# CLNE # todo explain how I got this data from running a sim and runing the cohort save data file \n",
    "\n",
    "VS_log_CLNE = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/previous_models/CLNE_files/cohort_saved_protein_count_data_10_09/validation_data/log_data/Log10_Schmidt_Comparison_startGen_2.csv')\n",
    "VW_log_CLNE = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/previous_models/CLNE_files/cohort_saved_protein_count_data_10_09/validation_data/log_data/Log10_Wisniewski_Comparison_startGen_2.csv')\n",
    "\n",
    "# CLClimNE\n",
    "# CHANGE THESE TO THE CORRECT PATH\n",
    "VS_log_CLClimNE = pd.read_csv(\n",
    "    '/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim3/Clim3d/cohort_saved_protein_count_data/validation_data/log_data/Log10_Schmidt_Comparison_startGen_2.csv')\n",
    "VW_log_CLClimNE = pd.read_csv(\n",
    "    '/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim3/Clim3d/cohort_saved_protein_count_data/validation_data/log_data/Log10_Wisniewski_Comparison_startGen_2.csv')"
   ],
   "id": "19ad1105a13104ac",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# specifiy the current sequence being used: \n",
    "current_sequence = 'CLClim3dNE'\n",
    "Clim_name = 'Clim3d'"
   ],
   "id": "f022b637c37be87a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Schmidt Validation Data Comparisons\n",
   "id": "1054105252719378"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot #1: Compare the Schmidt Validation Protein Counts for the CLNE and CLClim0NE models",
   "id": "6ecec8c713aad371"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "VS_log_CLClimNE",
   "id": "539c6abe2aee62fb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline(fig, df, name):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"Monomer ID\"]\n",
    "\n",
    "    # Add scatter trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=name))\n",
    "\n",
    "    # Compute trendline\n",
    "    z = np.polyfit(x, y, 1)\n",
    "    p = np.poly1d(z)\n",
    "    trendline_y = p(x)\n",
    "\n",
    "    # Add trendline trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline(fig, VS_log_CLNE, 'CLNE vs Schmidt')\n",
    "add_scatter_and_trendline(fig, VS_log_CLClimNE, (current_sequence +' vs Schmidt'))\n",
    "\n",
    "# Update layout\n",
    "fig.update_traces(marker_size=3)\n",
    "fig.update_layout(\n",
    "    title=f\"Schmidt Validation Protein Counts vs CLNE and {current_sequence}\",\n",
    "    xaxis_title=\"log10(Schmidt Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(CLNE, {current_sequence})\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600\n",
    ")\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "4f970538b2ed8d5b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot #1.a: Compare the Schmidt Validation Protein Counts for the CLNE and CLClim0NE models but with seaborn.",
   "id": "5de58468fc55f15c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import seaborn as sns\n",
    "\n",
    "# TODO: ask nora if this should be normalized by the total protein counts in the validation data? the way I did for cyrus's data?\n",
    "\n",
    "# Plot the data:\n",
    "sns.scatterplot(x=\"Log10 Validation Schmidt Counts\", y=\"Log10 Simulated Schmidt Counts\", data=VS_log_CLNE, s=5, color='lightseagreen', alpha=.8)\n",
    "sns.scatterplot(x=\"Log10 Validation Schmidt Counts\", y=\"Log10 Simulated Schmidt Counts\", data=VS_log_CLClimNE, s=5, color=\"orange\", alpha=.8)\n",
    "\n",
    "# find the max value of the plot for making a y=x line: \n",
    "max_vals = [max(VS_log_CLNE[\"Log10 Validation Schmidt Counts\"]), max(VS_log_CLNE[\"Log10 Simulated Schmidt Counts\"])]; max_val = round(max(max_vals))\n",
    "yxvals = range(0, max_val + 2) \n",
    "plt.plot(yxvals, yxvals, linewidth=.8, linestyle=\"dashed\", color=\"black\", alpha=.2); \n",
    "\n",
    "# plot specs\n",
    "plt.xlabel(\"log10(Schmidt Validation Protein Counts)\")\n",
    "plt.ylabel(f\"log10(Simulation Protein Counts)\")\n",
    "plt.axis('square')\n",
    "plt.title(\"Comparison between $\\t{E. coli}$ Protein Counts from Proteomics \\n Validation Data (Schmidt, et al., 2016) and \\nWhole-Cell Model (WCM) Simulation Data\")\n",
    "plt.legend([\"Original WCM (Macklin, et al., 2020)\", \"Current WCM (including half-lives assigned to values\\n from Gupta, et al., 2024)\",], bbox_to_anchor=(.5, -.35), loc='lower center')"
   ],
   "id": "effeb726fb6c6261",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "This is the same plot as above, but I normalized one set to the other. Not sure that it changed anything though.",
   "id": "49a4908680960b5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "1add663e2ede8271",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot #2: Compare the Schmidt Validation Protein Counts for the CLNE and CLClim0NE models (but filter values below log10(2))",
   "id": "b221de9dce47f12f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Plot the simulation protein counts against the validation counts, filtering out proteins with counts below log10(2) (these are likely to be subgenerational proteins) \n",
    "def add_scatter_and_trendline(fig, df, name):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"Monomer ID\"]\n",
    "\n",
    "    # Add scatter \n",
    "    fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=name))\n",
    "\n",
    "    # Compute trendline for data \n",
    "    z = np.polyfit(x, y, 1)\n",
    "    p = np.poly1d(z)\n",
    "    trendline_y = p(x)\n",
    "\n",
    "    # Add trendline \n",
    "    fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# filter out any counts below log10(2):\n",
    "VS_log_CLNE_filtered = VS_log_CLNE[(VS_log_CLNE[\"Log10 Validation Schmidt Counts\"] > 2) & (VS_log_CLNE[\"Log10 Simulated Schmidt Counts\"] > 2)]\n",
    "VS_log_CLClimNE_filtered = VS_log_CLClimNE[(VS_log_CLClimNE[\"Log10 Validation Schmidt Counts\"] > 2) & (VS_log_CLClimNE[\"Log10 Simulated Schmidt Counts\"] > 2)]\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline(fig, VS_log_CLNE_filtered, 'CLNE vs Schmidt')\n",
    "add_scatter_and_trendline(fig, VS_log_CLClimNE_filtered, (current_sequence +' vs Schmidt'))\n",
    "\n",
    "# Update layout\n",
    "fig.update_traces(marker_size=3)\n",
    "fig.update_layout(\n",
    "    title=f\"Schmidt Validation Protein Counts vs CLNE and {current_sequence} (filtered for counts > log10(2))\",\n",
    "    xaxis_title=\"log10(Schmidt Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(CLNE, {current_sequence})\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600\n",
    ")\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[2, 6], y=[2, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "98099a4996469b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Wisniewski Validation Data Comparisons",
   "id": "f64087bd0ff302f1"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot #3: Compare the Wisniewski Validation Protein Counts for the CLNE and CLClim0NE models",
   "id": "c4526b99f0ed95b4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline(fig, df, name):\n",
    "    x = df[\"Log10 Validation Wisniewski Counts\"]\n",
    "    y = df[\"Log10 Simulated Wisniewski Counts\"]\n",
    "    hovertext = df[\"Monomer ID\"]\n",
    "\n",
    "    # Add scatter trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=name))\n",
    "\n",
    "    # Compute trendline\n",
    "    z = np.polyfit(x, y, 1)\n",
    "    p = np.poly1d(z)\n",
    "    trendline_y = p(x)\n",
    "\n",
    "    # Add trendline trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline(fig, VW_log_CLNE, 'CLNE vs Wisniewski')\n",
    "add_scatter_and_trendline(fig, VW_log_CLClimNE, (current_sequence +' vs Wisniewski'))\n",
    "\n",
    "# Update layout\n",
    "fig.update_traces(marker_size=3)\n",
    "fig.update_layout(\n",
    "    title=f\"Wisniewski Validation Protein Counts vs CLNE and {current_sequence}\",\n",
    "    xaxis_title=\"log10(Wisniewski Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(CLNE, {current_sequence})\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600\n",
    ")\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "a4b2299f53ec6f7a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot #4: Compare the Wisniewski Validation Protein Counts for the CLNE and CLClim0NE models (but filter values below log10(2))",
   "id": "9102309ed6a88696"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline(fig, df, name):\n",
    "    x = df[\"Log10 Validation Wisniewski Counts\"]\n",
    "    y = df[\"Log10 Simulated Wisniewski Counts\"]\n",
    "    hovertext = df[\"Monomer ID\"]\n",
    "\n",
    "    # Add scatter trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=name))\n",
    "\n",
    "    # Compute trendline\n",
    "    z = np.polyfit(x, y, 1)\n",
    "    p = np.poly1d(z)\n",
    "    trendline_y = p(x)\n",
    "\n",
    "    # Add trendline trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "# filter out any counts below log10(2):\n",
    "VW_log_CLNE = VW_log_CLNE[(VW_log_CLNE[\"Log10 Validation Wisniewski Counts\"] > 2) & (VW_log_CLNE[\"Log10 Simulated Wisniewski Counts\"] > 2)]\n",
    "VW_log_CLClimNE = VW_log_CLClimNE[(VW_log_CLClimNE[\"Log10 Validation Wisniewski Counts\"] > 2) & (VW_log_CLClimNE[\"Log10 Simulated Wisniewski Counts\"] > 2)]\n",
    "add_scatter_and_trendline(fig, VW_log_CLNE, 'CLNE vs Wisniewski')\n",
    "add_scatter_and_trendline(fig, VW_log_CLClimNE, (current_sequence +' vs Wisniewski'))\n",
    "\n",
    "# Update layout\n",
    "fig.update_traces(marker_size=3)\n",
    "fig.update_layout(\n",
    "    title=f\"Wisniewski Validation Protein Counts vs CLNE and {current_sequence}\",\n",
    "    xaxis_title=\"log10(Wisniewski Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(CLNE, {current_sequence})\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600\n",
    ")\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[2, 6], y=[2, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "baeffe413203de21",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Plot of the proteases in the CLClimNE model",
   "id": "dbc29c32dcb69cb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# FUNCTIONS\n",
    "\n",
    "# get the gene IDs for each monomer IDs:\n",
    "def get_gene_ids_for_monomer_ids():\n",
    "    # code adapted from convert_to_flat.py\n",
    "    import io\n",
    "    from wholecell.io import tsv\n",
    "    from wholecell.utils.filepath import ROOT_PATH\n",
    "    RNAS_FILE = os.path.join(ROOT_PATH, 'reconstruction', 'ecoli', 'flat', 'rnas.tsv')\n",
    "    with io.open(RNAS_FILE, 'rb') as f:\n",
    "        reader = tsv.reader(f, delimiter='\\t')\n",
    "        headers = next(reader)\n",
    "        while headers[0].startswith('#'):\n",
    "            headers = next(reader)\n",
    "\n",
    "        gene_id_index = headers.index('gene_id')\n",
    "        protein_id_index = headers.index('monomer_ids')\n",
    "        monomer_ids_to_gene_ids = {}\n",
    "        for line in reader:\n",
    "            gene_id = line[gene_id_index]\n",
    "            protein_id = list(line[protein_id_index][2:-2].split('\", \"'))[0]\n",
    "            monomer_ids_to_gene_ids[protein_id] = gene_id\n",
    "\n",
    "        return monomer_ids_to_gene_ids\n",
    "    \n",
    "def get_gene_symbols_for_monomer_ids():\n",
    "    # code adapted from convert_to_flat.py\n",
    "    import io\n",
    "    from wholecell.io import tsv\n",
    "    from wholecell.utils.filepath import ROOT_PATH\n",
    "    RNAS_FILE = os.path.join(ROOT_PATH, 'reconstruction', 'ecoli', 'flat', 'rnas.tsv')\n",
    "    with io.open(RNAS_FILE, 'rb') as f:\n",
    "        reader = tsv.reader(f, delimiter='\\t')\n",
    "        headers = next(reader)\n",
    "        while headers[0].startswith('#'):\n",
    "            headers = next(reader)\n",
    "\n",
    "        gene_symbol_index = headers.index('common_name')\n",
    "        protein_id_index = headers.index('monomer_ids')\n",
    "        monomer_ids_to_gene_symbols = {}\n",
    "        for line in reader:\n",
    "            gene_symbol = line[gene_symbol_index]\n",
    "            protein_id = list(line[protein_id_index][2:-2].split('\", \"'))[0]\n",
    "            monomer_ids_to_gene_symbols[protein_id] = gene_symbol\n",
    "\n",
    "        return monomer_ids_to_gene_symbols\n",
    "    \n",
    "    \n",
    "# convert gene IDs to monomer IDs:\n",
    "def get_monomer_ids_for_gene_ids():\n",
    "    # code adapted from convert_to_flat.py\n",
    "    import io\n",
    "    from wholecell.io import tsv\n",
    "    from wholecell.utils.filepath import ROOT_PATH\n",
    "    RNAS_FILE = os.path.join(ROOT_PATH, 'reconstruction', 'ecoli', 'flat', 'rnas.tsv')\n",
    "    with io.open(RNAS_FILE, 'rb') as f:\n",
    "        reader = tsv.reader(f, delimiter='\\t')\n",
    "        headers = next(reader)\n",
    "        while headers[0].startswith('#'):\n",
    "            headers = next(reader)\n",
    "\n",
    "        gene_id_index = headers.index('gene_id')\n",
    "        protein_id_index = headers.index('monomer_ids')\n",
    "        gene_ids_to_monomer_ids = {}\n",
    "        for line in reader:\n",
    "            gene_id = line[gene_id_index]\n",
    "            protein_id = list(line[protein_id_index][2:-2].split('\", \"'))[0]\n",
    "            gene_ids_to_monomer_ids[gene_id] = protein_id\n",
    "\n",
    "        return gene_ids_to_monomer_ids\n",
    "    \n",
    "    \n",
    "# recall the genes defined as \"essential\" in the model, and return them as a list of gene names:\n",
    "def get_essential_genes():\n",
    "    # code adapted from convert_to_flat.py\n",
    "    import io\n",
    "    from wholecell.io import tsv\n",
    "    from wholecell.utils.filepath import ROOT_PATH\n",
    "    GENES_FILE = os.path.join(ROOT_PATH, 'validation', 'ecoli', 'flat', 'essential_genes.tsv')\n",
    "    with io.open(GENES_FILE, 'rb') as f:\n",
    "        reader = tsv.reader(f, delimiter='\\t')\n",
    "        headers = next(reader)\n",
    "        while headers[0].startswith('#'):\n",
    "            headers = next(reader)\n",
    "\n",
    "        gene_id_index = headers.index('FrameID')\n",
    "        essential_genes = []\n",
    "        for line in reader:\n",
    "            essential_gene = line[gene_id_index]\n",
    "            essential_genes.append(essential_gene)\n",
    "\n",
    "        return essential_genes\n",
    "\t\n",
    "\t\n",
    "# convert the data from the ParCa translation.py raw HL saves: \n",
    "def convert_HL_data(dataframe):\n",
    "    name_s = 'rate constant (s^1)' # name in seconds\n",
    "    name_m = 'half life (min)' # name in mins\n",
    "    dataframe.rename(columns={'Rate Constant': name_s}, inplace=True)\n",
    "    dataframe[name_m] = dataframe[name_s] * 60 * (1/np.log(2))\n",
    "    dataframe[name_m] = 1 / dataframe[name_m] # units of mins\n",
    "    return dataframe\n",
    "\n",
    "\n",
    "# convert each data source in the saved HL data file folders \n",
    "def convert_data_and_add_HL_source(dataframe, HL_source):\n",
    "    name_s = HL_source + ' rate constant (s^1)' # name in seconds\n",
    "    name_m = HL_source + ' half life (min)' # name in mins\n",
    "    dataframe.rename(columns={'Rate Constant': name_s}, inplace=True)\n",
    "    dataframe[name_m] = dataframe[name_s] * 60 * (1/np.log(2))\n",
    "    dataframe[name_m] = 1 / dataframe[name_m] # units of mins\n",
    "    return dataframe\n",
    "    \n"
   ],
   "id": "e3d560a76300ae28",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# load in the protease-substrate assignment datatable: \n",
    "protease_data = pd.read_excel('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/supplementary_data/Gupta_et_al_2024_ST2_protease_assignment_data.xlsx')\n",
    "\n",
    "# determine if any of the proteins assigned to a protease are considered an essential gene in the model: \n",
    "essential_substrate_gene_symbols = []\n",
    "for gene in get_essential_genes():\n",
    "    if get_gene_symbols_for_monomer_ids()[get_monomer_ids_for_gene_ids()[gene]] in protease_data['Gene name'].values:\n",
    "        essential_substrate_gene_symbols.append(get_gene_symbols_for_monomer_ids()[get_monomer_ids_for_gene_ids()[gene]])\n",
    "        \n",
    "print(\"There are \" + str(len(protease_data['Gene name'].values)) + \" proteins assigned to a protease in the Gupta et al., 2024 data.\")\n",
    "print(\"Essential Genes in Model that are Assigned to a Protease (\" +str(len(essential_substrate_gene_symbols))+\"): \", essential_substrate_gene_symbols)"
   ],
   "id": "99936ff7c9dc3785",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# determine which monomer IDs are protein substrates of the proteases in the CLClim0NE model:\n",
    "monomer_ids = VS_log_CLClimNE['Monomer ID']\n",
    "monomer_ids = monomer_ids.str.slice(0, -3)\n",
    "\n",
    "CLClim0NE_protease_substrate_gene_symbols = []\n",
    "CLClim0NE_protease_substrate_monomer_ids = []\n",
    "for monomer_id in monomer_ids:\n",
    "\tmonomer = monomer_id\n",
    "\tgene_symbol = get_gene_symbols_for_monomer_ids()[monomer]\n",
    "\tif gene_symbol in protease_data['Gene name'].values:\n",
    "\t\tCLClim0NE_protease_substrate_gene_symbols.append(gene_symbol)\n",
    "\t\tCLClim0NE_protease_substrate_monomer_ids.append(monomer)\n",
    "        \n",
    "print(\"Total Protease Substrates in VS CLClimNE (\" +str(len(CLClim0NE_protease_substrate_gene_symbols))+\"): \", CLClim0NE_protease_substrate_gene_symbols)\n",
    "\n",
    "VS_log_CLClimNE['gene symbol'] = VS_log_CLClimNE['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]]) # this is to add to the hovertext\n",
    "# if a monomer in CLClim0NE_protease_substrate_gene_symbols is found in monomer_ids, remove them from Clim0_data and add it to a new dataframe: \n",
    "CLClim0NE_protease_substrate_idxs = monomer_ids.isin(CLClim0NE_protease_substrate_monomer_ids)\n",
    "CLClim0NE_protease_substrates = VS_log_CLClimNE[CLClim0NE_protease_substrate_idxs].copy()\n",
    "CLClim0NE_protease_substrates['HL Source'] = f\"{current_sequence} (protease substrate)\"\n",
    "CLClim0NE_VS_log_data_remaining = VS_log_CLClimNE[~CLClim0NE_protease_substrate_idxs].copy()\n",
    "# Not all the proteases match to model proteins here becuase the Schmidt data has less proteins overall! so thtat is why the numbers are different."
   ],
   "id": "f6148d3ac78ec354",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "#Todo: this might need to be updated with the protease names altered to match the model's nomenclature.\n",
    "\n",
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline(fig, df, name, is_protease_substrate=False):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"gene symbol\"]\n",
    "\t\n",
    "    # Properly indented if statement\n",
    "    if is_protease_substrate == True:\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', \n",
    "                                  marker_size=7, marker=dict(color=\"orange\", symbol='star'), \n",
    "                                  name=name))\n",
    "    else: \n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker_size=4,\n",
    "                                 marker=dict(color=\"lightseagreen\"),  name=name))\n",
    "\t\t\n",
    "        # Compute trendline\n",
    "        z = np.polyfit(x, y, 1)\n",
    "        p = np.poly1d(z)\n",
    "        trendline_y = p(x)\n",
    "    \n",
    "        # Add trendline trace\n",
    "        fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', line=dict(color=\"salmon\"), name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline(fig, CLClim0NE_VS_log_data_remaining, (current_sequence + ' vs Schmidt'))\n",
    "add_scatter_and_trendline(fig, CLClim0NE_protease_substrates, (current_sequence + '(protease substrates) vs Schmidt'), True)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Schmidt Validation Protein Counts vs {current_sequence}\",\n",
    "    xaxis_title=\"log10(Schmidt Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10({current_sequence})\",\n",
    "    autosize=False,\n",
    "    width=1200,\n",
    "    height=1000\n",
    ")\n",
    "\n",
    "# Add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=dict(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"))\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "5e3fa0516e72b9e9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline_proposal(fig, df, name, is_protease_substrate=False):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"gene symbol\"]\n",
    "\t\n",
    "    # Properly indented if statement\n",
    "    if is_protease_substrate == True:\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', \n",
    "                                  marker_size=7, marker=dict(color=\"orange\", symbol='star'), \n",
    "                                  name=name))\n",
    "    else: \n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker_size=4,\n",
    "                                 marker=dict(color=\"lightseagreen\", opacity=.5),  name=name))\n",
    "\t\t\n",
    "        # Compute trendline\n",
    "        z = np.polyfit(x, y, 1)\n",
    "        p = np.poly1d(z)\n",
    "        trendline_y = p(x)\n",
    "    \n",
    "        # Add trendline trace\n",
    "        fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', line=dict(color=\"salmon\"), name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline_proposal(fig, CLClim0NE_VS_log_data_remaining, (current_sequence + ' vs Schmidt'))\n",
    "add_scatter_and_trendline_proposal(fig, CLClim0NE_protease_substrates, (current_sequence + '(protease substrates) vs Schmidt'), True)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"WCM Protein Counts (incorperating HLs from Gupta, et al., 2024) <br>vs Validation Protein Counts (Schmidt, et al., 2016)\",\n",
    "    xaxis_title=\"log10(Schmidt Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(WCM incorperating Clim HLs Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=1000,\n",
    "    height=700\n",
    ")\n",
    "\n",
    "# Add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=dict(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"))\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "9c9a006558bd8c4f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Make the above plot but with essential genes as the stars instead",
   "id": "475cf9d848a6e6d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# determine the essential genes in VS CLClimNE\n",
    "VS_log_CLClimNE['gene symbol'] = VS_log_CLClimNE['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]]) \n",
    "gene_ids = VS_log_CLClimNE['gene symbol']\n",
    "        \n",
    "CLClim0NE_essential_gene_symbols = []\n",
    "CLClim0NE_essential_gene_monomer_ids = []\n",
    "for gene in get_essential_genes():\n",
    "    gene_symbol = get_gene_symbols_for_monomer_ids()[get_monomer_ids_for_gene_ids()[gene]]\n",
    "    if gene_symbol in VS_log_CLClimNE['gene symbol'].values:\n",
    "        CLClim0NE_essential_gene_symbols.append(gene_symbol)\n",
    "        CLClim0NE_essential_gene_monomer_ids.append(get_monomer_ids_for_gene_ids()[gene])\n",
    "        \n",
    "print(\"Total Essential Genes in VS CLClimNE (\" +str(len(CLClim0NE_essential_gene_symbols))+\"): \", CLClim0NE_essential_gene_symbols)"
   ],
   "id": "8d0ada658cc399b5",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "# if a monomer in CLClim0NE_essential_gene_monomer_ids is found in monomer_ids, remove them from Clim0_data and add it to a new dataframe: \n",
    "CLClim0NE_EG_idxs = monomer_ids.isin(CLClim0NE_essential_gene_monomer_ids)\n",
    "CLClim0NE_EGs = VS_log_CLClimNE[CLClim0NE_EG_idxs].copy()\n",
    "CLClim0NE_EGs['HL Source'] = f\"{current_sequence} (essential gene)\"\n",
    "CLClim0NE_VS_log_data_remaining = VS_log_CLClimNE[~CLClim0NE_EG_idxs].copy()"
   ],
   "id": "4c6d5150b955b05",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline_EGs(fig, df, name, is_protease_substrate=False):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"gene symbol\"]\n",
    "\t\n",
    "    # Properly indented if statement\n",
    "    if is_protease_substrate == True:\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', \n",
    "                                  marker_size=7, marker=dict(color=\"orange\", symbol='star'), \n",
    "                                  name=name))\n",
    "    else: \n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker_size=4,\n",
    "                                 marker=dict(color=\"lightseagreen\", opacity=.5),  name=name))\n",
    "\t\t\n",
    "        # Compute trendline\n",
    "        z = np.polyfit(x, y, 1)\n",
    "        p = np.poly1d(z)\n",
    "        trendline_y = p(x)\n",
    "    \n",
    "        # Add trendline trace\n",
    "        fig.add_trace(go.Scatter(x=x, y=trendline_y, mode='lines', line=dict(color=\"salmon\"), name=f'{name} Trendline: {p}'))\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline_EGs(fig, CLClim0NE_VS_log_data_remaining, (current_sequence + ' vs Schmidt'))\n",
    "add_scatter_and_trendline_EGs(fig, CLClim0NE_EGs, (current_sequence + '(essential genes) vs Schmidt'), True)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"WCM Protein Counts (incorperating HLs from Gupta, et al., 2024) <br>vs Validation Protein Counts (Schmidt, et al., 2016)\",\n",
    "    xaxis_title=\"log10(Schmidt Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(WCM incorperating Clim HLs Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=1000,\n",
    "    height=700\n",
    ")\n",
    "\n",
    "# Add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=dict(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"))\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "78fe844d088fe9fa",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# do the matplotlib version of the above plot: \n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Function to create scatter plot and add trendline for Matplotlib\n",
    "def add_scatter_and_trendline_EGs(ax, df, name, is_protease_substrate=False):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"gene symbol\"]  # Note: Matplotlib doesn't support hover text\n",
    "\n",
    "    if is_protease_substrate:\n",
    "        ax.scatter(x, y, color=\"orange\", s=70, marker='*', label=f'Essential Protein', edgecolor='none', alpha=.9, zorder=2)\n",
    "    else:\n",
    "        ax.scatter(x, y, color=\"lightseagreen\", edgecolor='none', s=20, alpha=0.3, zorder=1)\n",
    "        \n",
    "# Create figure and axis\n",
    "fig, ax = plt.subplots(figsize=(10, 7), dpi=200)\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline_EGs(ax, CLClim0NE_VS_log_data_remaining, f'{current_sequence} vs Schmidt')\n",
    "add_scatter_and_trendline_EGs(ax, CLClim0NE_EGs, f'{current_sequence} (essential genes) vs Schmidt', True)\n",
    "\n",
    "# Add y=x reference line\n",
    "ax.plot([-.250, 6], [-.250, 6], color=\"black\", linestyle=\"--\", linewidth=1, alpha=0.2,  zorder=0)\n",
    "\n",
    "# Add title and axis labels\n",
    "ax.set_title(\"Updated WCM Protein Counts vs Validation Protein Counts\", fontsize=14)\n",
    "ax.set_xlabel(\"log10(Schmidt et al. Validation Protein Counts)\", fontsize=12)\n",
    "ax.set_ylabel(\"log10(WCM w/ Gupta et al. Clim HLs Protein Counts)\", fontsize=12)\n",
    "ax.set_xlim(-.250, 6)\n",
    "ax.set_ylim(-.250, 6)\n",
    "# make the axes square\n",
    "ax.set_aspect('equal', adjustable='box')\n",
    "\n",
    "# Add legend\n",
    "ax.legend()\n",
    "\n",
    "# Show plot\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# determine the R squared value "
   ],
   "id": "c17fb152e56a069d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sklearn.metrics import r2_score\n",
    "\n",
    "# reconnect the data: \n",
    "CLClim0NE_together = pd.concat([CLClim0NE_VS_log_data_remaining, CLClim0NE_EGs])\n",
    "\n",
    "# rsquared = r2_score( true, predicted)\n",
    "r_squared = r2_score(CLClim0NE_together[\"Log10 Validation Schmidt Counts\"], CLClim0NE_together[\"Log10 Simulated Schmidt Counts\"])\n",
    "r_squared\n"
   ],
   "id": "c669d81356700b93",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "CLClim0NE_together",
   "id": "eea390dd75c703b6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# determine the R squared value separately for the validaiton counts greater than 30 and less than 30 \n",
    "CLClim0NE_together_30_above = CLClim0NE_together[CLClim0NE_together[\"Log10 Validation Schmidt Counts\"] >= np.log10(30+1)]\n",
    "CLClim0NE_together_30_above\n"
   ],
   "id": "c1d50fda61e17e12",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "r_squared_30_above = r2_score(CLClim0NE_together_30_above[\"Log10 Validation Schmidt Counts\"], CLClim0NE_together_30_above[\"Log10 Simulated Schmidt Counts\"])\n",
    "\n",
    "# calculate the R squared value for the validation counts less than 30:\n",
    "CLClim0NE_together_30_below = CLClim0NE_together[(CLClim0NE_together[\"Log10 Validation Schmidt Counts\"] < np.log(30+1))]\n",
    "r_squared_30_below = r2_score(CLClim0NE_together_30_below[\"Log10 Validation Schmidt Counts\"], CLClim0NE_together_30_below[\"Log10 Simulated Schmidt Counts\"])\n",
    "\n",
    "r_squared_30_above, r_squared_30_below"
   ],
   "id": "b274dbbf637474f1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Proposal plots",
   "id": "35610c8714e76bcd"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Different plots \n",
    "\n",
    "Plot a: comparison of the original model to the new CLClime0NE model\n",
    "\n",
    "Plot b: plot of CLClime0NE model (with the protease substrates highlighted) plotted against Schmidt data\n",
    "\n",
    "Plot c: plot of the half-lives assigned to Clim0 valuesin CLClime0NE model (with the protease substrates highlighted) that were originally assigned to 10 hours by NE in the original model\n",
    "\n",
    "Plot d: plot of the half-lives assigned to Clim0 values in CLClime0NE model (with the protease substrates highlighted) that were originally assigned to 2 mins in the NE in the original model"
   ],
   "id": "221f42ff3f93b0d2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# first find the monomer IDs that are substrates of the proteases in the CLClim0NE model:\n",
    "protease_data = pd.read_excel('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/supplementary_data/Gupta_et_al_2024_ST2_protease_assignment_data.xlsx')\n",
    "\n",
    "# read in the unfiltered data: \n",
    "CLClimNE_log_data_with_proteases = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim1/CLClim1NE/CLCLim1NE_cohort_saved_protein_count_data/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "# add the gene symbols to the dataframe:\n",
    "CLClimNE_log_data_with_proteases['gene symbol'] = CLClimNE_log_data_with_proteases['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]])\n",
    "CLClimNE_log_data_with_proteases['protease assignement'] = \"NA\"\n",
    "CLClimNE_log_data_with_proteases"
   ],
   "id": "f4799bad34c38042",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# determine which monomer IDs are protein substrates of the proteases in the CLClim0NE model:\n",
    "gene_ids = CLClimNE_log_data_with_proteases['gene symbol']\n",
    "\n",
    "CLClim0NE_full_protease_substrate_gene_symbols = []\n",
    "for gene_id in gene_ids:\n",
    "    if gene_id in protease_data['Gene name'].values:\n",
    "        CLClim0NE_full_protease_substrate_gene_symbols.append(gene_id)\n",
    "        protease_row = protease_data[protease_data['Gene name'] == gene_id]\n",
    "        protease = protease_row['Protease assignment'].values[0]  \n",
    "        CLClimNE_log_data_with_proteases.loc[CLClimNE_log_data_with_proteases['gene symbol'] == gene_id, 'protease assignement'] = protease\n",
    "\n",
    "print(\"Total Protease Substrates in CLClim0NE (\" +str(len(CLClim0NE_full_protease_substrate_gene_symbols))+\"): \", CLClim0NE_full_protease_substrate_gene_symbols)\n",
    "# Differences in (308 to 236) numbers are likely due the the fact that the strain of E. coli in the paper might differ from the one in our model. \n",
    "CLClimNE_log_data_with_proteases"
   ],
   "id": "a8ceeda8c4039106",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# figure out which proteins were assigned to lon in the model: \n",
    "lon_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'Lon only']\n",
    "lon_proteins\n",
    "\n",
    "# figure out which proteins were assigned to clp in the model:\n",
    "clp_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'ClpP only']\n",
    "\n",
    "hslv_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'HslV only']\n",
    "\n",
    "# figure out which proteins are additive in the model: \n",
    "additive_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'Additive']\n",
    "\n",
    "# figure out which proteins are redundant in the model: \n",
    "redundant_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'Redundant']\n",
    "\n",
    "# figure out which proteins are assinged as \"unexplained\" in the model: \n",
    "unexplained_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'Actively degrading in Tripple KO']\n",
    "\n",
    "# figure out which proteins were not assigned to a protease in the model:\n",
    "no_protease_proteins = CLClimNE_log_data_with_proteases[CLClimNE_log_data_with_proteases['protease assignement'] == 'NA']\n",
    "\n",
    "# todo: note that these numbers will not add up to those in the graph below bc some proteins in the CLClim0NE model are not in the CLNE model \n",
    "#298 total \n",
    "print(\"Lon Proteins(\" +str(len(lon_proteins))+\"): \", list(lon_proteins['gene symbol'])) #14\n",
    "print(\"ClpP Proteins(\" +str(len(clp_proteins))+\"): \", list(clp_proteins['gene symbol'])) #63\n",
    "print(\"HslV Proteins(\" +str(len(hslv_proteins))+\"): \", list(hslv_proteins['gene symbol'])) #1\n",
    "print(\"Additive Proteins(\" +str(len(additive_proteins))+\"): \", list(additive_proteins['gene symbol'])) #79\n",
    "print(\"Redundant Proteins(\" +str(len(redundant_proteins))+\"): \", list(redundant_proteins['gene symbol'])) #39\n",
    "print(\"Unexplained Proteins(\" +str(len(unexplained_proteins))+\"): \", list(unexplained_proteins['gene symbol'])) #102\n",
    "#print(\"No Protease Proteins(\" +str(len(no_protease_proteins))+\"): \", list(no_protease_proteins['gene symbol'])) # 4012\n"
   ],
   "id": "96bdfec11b3b4d0d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Pre plot: try to split up by protease assignment",
   "id": "dac3e7a3a7099c8e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# read in the unfiltered data: \n",
    "CLClimNE_log_data_with_proteases"
   ],
   "id": "51bc0c7775c6d525",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "#CLNE_log_data_no_idx['Monomer ID'].isin(CLClimNE_log_data_with_proteases_no_idx['Monomer ID']).sum()\n",
   "id": "77b642dbb34cc510",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "#CLClimNE_log_data_with_proteases_no_idx[~CLClimNE_log_data_with_proteases_no_idx['Monomer ID'].isin(CLNE_log_data_no_idx['Monomer ID'])]\n",
   "id": "38a0c6be1a8ab75a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "## Create scatter plot of CLNE counts to CLClim0NE counts, but color the protiens based on their protease assignment in the CLClim0NE model\n",
    "# create the plot!\n",
    "# TODO: do I need to normalize the data to ? and how do I get the plot to be an exact square? and the figure title centered?\n",
    "def CLNE_CLClimNE_plot_protease(fig, dataframe, protease=0):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    "    hovertext = dataframe[\"gene symbol\"]\n",
    " \n",
    "    # Add scatter trace\n",
    "    if protease == 0: # no assignement\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"No Protease ({np.shape(dataframe)[0]})\", marker=dict(color='orange', size=3, opacity=.5)))\n",
    "\n",
    "    elif protease == 1: # no explaination\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Unexplained ({np.shape(dataframe)[0]})\", marker=dict(color='deeppink', size=3, opacity=.5)))\n",
    "        \n",
    "    elif protease == 2: # redundant\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Redundant ({np.shape(dataframe)[0]})\", marker=dict(color='lightseagreen', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 3: # additive\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Additive ({np.shape(dataframe)[0]})\", marker=dict(color='blue', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 4: # lon\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Lon ({np.shape(dataframe)[0]})\", marker=dict(color='red', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 5: # clp\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"ClpP ({np.shape(dataframe)[0]})\", marker=dict(color='green', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 6: # hslv\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Hslv ({np.shape(dataframe)[0]})\", marker=dict(color='yellow', size=5, opacity=.9)))\n",
    "\n",
    "\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "del CLClimNE_log_data_with_proteases\n",
    "# first find the monomer IDs that are substrates of the proteases in the CLClim0NE model:\n",
    "protease_data = pd.read_excel('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/supplementary_data/Gupta_et_al_2024_ST2_protease_assignment_data.xlsx')\n",
    "\n",
    "# read in the unfiltered data: \n",
    "CLClimNE_log_data_with_proteases = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim1/CLClim1NE/CLCLim1NE_cohort_saved_protein_count_data/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "CLNE_log_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/previous_models/CLNE_files/cohort_saved_protein_count_data_10_09/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "# merge the two dataframes:\n",
    "CLNE_log_data = CLNE_log_data.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLNE Protein Counts\"})\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLClim0NE Protein Counts\"})\n",
    "CLNE_log_data = CLNE_log_data[[\"Monomer ID\", \"Log10 CLNE Protein Counts\"]]\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases[[\"Monomer ID\", \"Log10 CLClim0NE Protein Counts\"]]\n",
    "CLNE_log_data = CLNE_log_data.set_index(\"Monomer ID\")\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases.set_index(\"Monomer ID\")\n",
    "CLNE_CLClimNE_log_data_proteases = CLNE_log_data.join(CLClimNE_log_data_with_proteases, how=\"inner\")\n",
    "CLNE_CLClimNE_log_data_proteases = CLNE_CLClimNE_log_data_proteases.reset_index()\n",
    "\n",
    "# add the gene symbols to the dataframe:\n",
    "CLNE_CLClimNE_log_data_proteases['gene symbol'] = CLNE_CLClimNE_log_data_proteases['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]])\n",
    "CLNE_CLClimNE_log_data_proteases['protease assignement'] = \"NA\"\n",
    "\n",
    "# determine which monomer IDs are protein substrates of the proteases in the CLClim0NE model:\n",
    "gene_ids = CLNE_CLClimNE_log_data_proteases['gene symbol']\n",
    "for gene_id in gene_ids:\n",
    "    if gene_id in protease_data['Gene name'].values:\n",
    "        protease_row = protease_data[protease_data['Gene name'] == gene_id]\n",
    "        protease = protease_row['Protease assignment'].values[0]  # Extract the protease assignment\n",
    "        # change the protease assignment for the gene id in the dataframe:\n",
    "        CLNE_CLClimNE_log_data_proteases.loc[CLNE_CLClimNE_log_data_proteases['gene symbol'] == gene_id, 'protease assignement'] = protease\n",
    "\n",
    "# make a copy of the CLClimNE_log_data_with_proteases data:\n",
    "CLClimNE_log_data_with_Lon_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Clp_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_HslV_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Additive_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Redundant_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Unexplained_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_no_protease_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "\n",
    "# remove the proteins that were not assigned to the protease in the model:\n",
    "CLClimNE_log_data_with_Lon_proteases = CLClimNE_log_data_with_Lon_proteases[CLClimNE_log_data_with_Lon_proteases['protease assignement'] == 'Lon only']\n",
    "CLClimNE_log_data_with_Clp_proteases = CLClimNE_log_data_with_Clp_proteases[CLClimNE_log_data_with_Clp_proteases['protease assignement'] == 'ClpP only']\n",
    "CLClimNE_log_data_with_HslV_proteases = CLClimNE_log_data_with_HslV_proteases[CLClimNE_log_data_with_HslV_proteases['protease assignement'] == 'HslV only']\n",
    "CLClimNE_log_data_with_Additive_proteases = CLClimNE_log_data_with_Additive_proteases[CLClimNE_log_data_with_Additive_proteases['protease assignement'] == 'Additive']\n",
    "CLClimNE_log_data_with_Redundant_proteases = CLClimNE_log_data_with_Redundant_proteases[CLClimNE_log_data_with_Redundant_proteases['protease assignement'] == 'Redundant']\n",
    "CLClimNE_log_data_with_Unexplained_proteases = CLClimNE_log_data_with_Unexplained_proteases[CLClimNE_log_data_with_Unexplained_proteases['protease assignement'] == 'Actively degrading in Tripple KO']\n",
    "CLClimNE_log_data_with_no_protease_proteases = CLClimNE_log_data_with_no_protease_proteases[CLClimNE_log_data_with_no_protease_proteases['protease assignement'] == 'NA']\n",
    "\n",
    "# now plot!\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_no_protease_proteases, 0)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Unexplained_proteases, 1)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Redundant_proteases, 2)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Additive_proteases, 3)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Lon_proteases, 4)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Clp_proteases, 5)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_HslV_proteases, 6)\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "    line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "    name=\"y=x\"));\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Protein Counts from the New WCM plotted <br>against those from the Original WCM\",\n",
    "    xaxis_title=\"log10(Orginal WCM Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600, \n",
    "    showlegend=True,\n",
    "    legend_title=\"Protease Assignments\")\n",
    "\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "f965b76c3ff6b442",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "np.shape(CLClimNE_log_data_with_proteases)",
   "id": "e87963e13a977448",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "np.shape(CLNE_log_data)",
   "id": "d4ea55090dfd022d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "make the smae plot as above, but only using the priority substates!",
   "id": "bcfe1737078ab01a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "## Create scatter plot of CLNE counts to CLClim0NE counts, but color the protiens based on their protease assignment in the CLClim0NE model\n",
    "# create the plot!\n",
    "# TODO: do I need to normalize the data? and how do I get the plot to be an exact square? and the figure title centered?\n",
    "def CLNE_CLClimNE_plot_protease(fig, dataframe, protease=0):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    "    hovertext = dataframe[\"gene symbol\"]\n",
    " \n",
    "    # Add scatter trace\n",
    "    if protease == 0: # no assignement\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"No Protease ({np.shape(dataframe)[0]})\", marker=dict(color='orange', size=3, opacity=.5)))\n",
    "\n",
    "    elif protease == 1: # no explaination\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Unexplained ({np.shape(dataframe)[0]})\", marker=dict(color='deeppink', size=3, opacity=.5)))\n",
    "        \n",
    "    elif protease == 2: # redundant\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Redundant ({np.shape(dataframe)[0]})\", marker=dict(color='lightseagreen', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 3: # additive\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Additive ({np.shape(dataframe)[0]})\", marker=dict(color='blue', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 4: # lon\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Lon ({np.shape(dataframe)[0]})\", marker=dict(color='red', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 5: # clp\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"ClpP ({np.shape(dataframe)[0]})\", marker=dict(color='green', size=5, opacity=.9)))\n",
    "        \n",
    "    elif protease == 6: # hslv\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', name=f\"Hslv ({np.shape(dataframe)[0]})\", marker=dict(color='yellow', size=5, opacity=.9)))\n",
    "\n",
    "\n",
    "\n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "del CLClimNE_log_data_with_proteases\n",
    "# first find the monomer IDs that are substrates of the proteases in the CLClim0NE model:\n",
    "priority_protease_data = pd.read_csv('/Users/miagrahn/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/supplementary_data/priority_protease_assignments_0.tsv')\n",
    "\n",
    "# read in the unfiltered data: \n",
    "CLClimNE_log_data_with_proteases = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim1/CLClim1NE/CLCLim1NE_cohort_saved_protein_count_data/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "CLNE_log_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/previous_models/CLNE_files/cohort_saved_protein_count_data_10_09/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "\n",
    "# merge the two dataframes:\n",
    "CLNE_log_data = CLNE_log_data.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLNE Protein Counts\"})\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLClim0NE Protein Counts\"})\n",
    "CLNE_log_data = CLNE_log_data[[\"Monomer ID\", \"Log10 CLNE Protein Counts\"]]\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases[[\"Monomer ID\", \"Log10 CLClim0NE Protein Counts\"]]\n",
    "CLNE_log_data = CLNE_log_data.set_index(\"Monomer ID\")\n",
    "CLClimNE_log_data_with_proteases = CLClimNE_log_data_with_proteases.set_index(\"Monomer ID\")\n",
    "CLNE_CLClimNE_log_data_proteases = CLNE_log_data.join(CLClimNE_log_data_with_proteases, how=\"inner\")\n",
    "CLNE_CLClimNE_log_data_proteases = CLNE_CLClimNE_log_data_proteases.reset_index()\n",
    "\n",
    "# add the gene symbols to the dataframe:\n",
    "CLNE_CLClimNE_log_data_proteases['gene symbol'] = CLNE_CLClimNE_log_data_proteases['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]])\n",
    "CLNE_CLClimNE_log_data_proteases['protease assignement'] = \"NA\"\n",
    "\n",
    "# determine which monomer IDs are protein substrates of the proteases in the CLClim0NE model:\n",
    "gene_ids = CLNE_CLClimNE_log_data_proteases['gene symbol']\n",
    "for gene_id in gene_ids:\n",
    "    if gene_id in priority_protease_data['Gene name'].values:\n",
    "        protease_row = priority_protease_data[priority_protease_data['Gene name'] == gene_id]\n",
    "        protease = protease_row['Protease assignment'].values[0]  # Extract the protease assignment\n",
    "        # change the protease assignment for the gene id in the dataframe:\n",
    "        CLNE_CLClimNE_log_data_proteases.loc[CLNE_CLClimNE_log_data_proteases['gene symbol'] == gene_id, 'protease assignement'] = protease\n",
    "\n",
    "# make a copy of the CLClimNE_log_data_with_proteases data:\n",
    "CLClimNE_log_data_with_Lon_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Clp_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_HslV_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Additive_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Redundant_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_Unexplained_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "CLClimNE_log_data_with_no_protease_proteases = CLNE_CLClimNE_log_data_proteases.copy()\n",
    "\n",
    "# remove the proteins that were not assigned to the protease in the model:\n",
    "CLClimNE_log_data_with_Lon_proteases = CLClimNE_log_data_with_Lon_proteases[CLClimNE_log_data_with_Lon_proteases['protease assignement'] == 'Lon only']\n",
    "CLClimNE_log_data_with_Clp_proteases = CLClimNE_log_data_with_Clp_proteases[CLClimNE_log_data_with_Clp_proteases['protease assignement'] == 'ClpP only']\n",
    "CLClimNE_log_data_with_HslV_proteases = CLClimNE_log_data_with_HslV_proteases[CLClimNE_log_data_with_HslV_proteases['protease assignement'] == 'HslV only']\n",
    "CLClimNE_log_data_with_Additive_proteases = CLClimNE_log_data_with_Additive_proteases[CLClimNE_log_data_with_Additive_proteases['protease assignement'] == 'Additive']\n",
    "CLClimNE_log_data_with_Redundant_proteases = CLClimNE_log_data_with_Redundant_proteases[CLClimNE_log_data_with_Redundant_proteases['protease assignement'] == 'Redundant']\n",
    "CLClimNE_log_data_with_Unexplained_proteases = CLClimNE_log_data_with_Unexplained_proteases[CLClimNE_log_data_with_Unexplained_proteases['protease assignement'] == 'Actively degrading in Tripple KO']\n",
    "CLClimNE_log_data_with_no_protease_proteases = CLClimNE_log_data_with_no_protease_proteases[CLClimNE_log_data_with_no_protease_proteases['protease assignement'] == 'NA']\n",
    "\n",
    "# now plot!\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_no_protease_proteases, 0)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Unexplained_proteases, 1)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Redundant_proteases, 2)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Additive_proteases, 3)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Lon_proteases, 4)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_Clp_proteases, 5)\n",
    "CLNE_CLClimNE_plot_protease(fig, CLClimNE_log_data_with_HslV_proteases, 6)\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "    line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "    name=\"y=x\"));\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Protein Counts from the New WCM plotted <br>against those from the Original WCM\",\n",
    "    xaxis_title=\"log10(Orginal WCM Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=900,\n",
    "    height=600, \n",
    "    showlegend=True,\n",
    "    legend_title=\"Protease Assignments\")\n",
    "\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "4444264f29f7d951",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# determine which of the 80 priority substrates are considered essential genes: \n",
    "essential_genes = pd.read_csv('~/wcEcoli/analysis/ecoli_essential_genes.csv')\n"
   ],
   "id": "b6bbd85ed3cf343e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot A: comparison of the original model to the new CLClime0NE model",
   "id": "bce1dd54e16715c2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# create a data frame that includes the log counts for the CLNE and CLClim0NE models:\n",
    "# todo determine if I should use the validation data numbers still? or should I use the filtred and/or unfiltered data? and should I normalize one to the other? \n",
    "CLClimNE_log_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim1/CLClim1NE/CLCLim1NE_cohort_saved_protein_count_data/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "CLNE_log_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/previous_models/CLNE_files/cohort_saved_protein_count_data_10_09/unfiltered_data/log_data/LogAvgProteinCounts_startGen_2.csv')\n",
    "\n",
    "# merge the two dataframes:\n",
    "CLNE_log_data = CLNE_log_data.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLNE Protein Counts\"})\n",
    "CLClimNE_log_data = CLClimNE_log_data.rename(columns={\"Log10 Average Protein Count\": \"Log10 CLClim0NE Protein Counts\"})\n",
    "CLNE_log_data = CLNE_log_data[[\"Monomer ID\", \"Log10 CLNE Protein Counts\"]]\n",
    "CLClimNE_log_data = CLClimNE_log_data[[\"Monomer ID\", \"Log10 CLClim0NE Protein Counts\"]]\n",
    "CLNE_log_data = CLNE_log_data.set_index(\"Monomer ID\")\n",
    "CLClimNE_log_data = CLClimNE_log_data.set_index(\"Monomer ID\")\n",
    "CLNE_CLClimNE_log_data = CLNE_log_data.join(CLClimNE_log_data, how=\"inner\")\n",
    "CLNE_CLClimNE_log_data = CLNE_CLClimNE_log_data.reset_index()\n",
    "\n",
    "# add the gene symbols to the dataframe:\n",
    "CLNE_CLClimNE_log_data['gene symbol'] = CLNE_CLClimNE_log_data['Monomer ID'].apply(lambda x: get_gene_symbols_for_monomer_ids()[x[:-3]])\n",
    "\n",
    "CLNE_CLClimNE_log_data "
   ],
   "id": "ce91d80270bff025",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# create the plot!\n",
    "# TODO: do I need to normalize the data? and how do I get the plot to be an exact square? and the figure title centered?\n",
    "def CLNE_CLClimNE_plot(fig, dataframe):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    " \n",
    "    # Add scatter trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=y, mode='markers', marker=dict(color='lightseagreen', size=3, opacity=.5)))\n",
    "\n",
    "    # add a y=x line\n",
    "    fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "    \n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "CLNE_CLClimNE_plot(fig, CLNE_CLClimNE_log_data)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Protein Counts from the New WCM plotted <br>against those from the Original WCM\",\n",
    "    xaxis_title=\"log10(Orginal WCM Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600, \n",
    "    showlegend=False)\n",
    "\n",
    "\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "29df3479406c0a15",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "a.1: color by change in stability: ",
   "id": "1b1697253b3a11f6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "cf66963bbf9657d6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "## Create scatter plot of CLNE counts to CLClim0NE counts, but color the protiens based on how their stability changed\n",
    "# create the plot!\n",
    "def CLNE_CLClimNE_plot_stability(fig, dataframe, stability=0):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    "    hovertext = dataframe[\"gene symbol\"]\n",
    " \n",
    "    # Add scatter trace\n",
    "    if stability == 0: # same stability\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker=dict(color='orange', size=3, opacity=.5)))\n",
    "\n",
    "    elif stability == 1: # less stable\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker=dict(color='deeppink', size=3, opacity=.5)))\n",
    "        \n",
    "    elif stability == 2: # more stable\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker=dict(color='lightseagreen', size=5, opacity=.9)))\n",
    "\n",
    "    \n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# load in the stability file from the CLClim0NE model:\n",
    "Clim_stability_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim0_using_biorxiv_2023_data/CLClim0NE/CLClim0NE_stability_change.csv')\n",
    "stability_changes = Clim_stability_data[[\"Protein ID\", \"Stability\"]]\n",
    "stability_changes = stability_changes.rename(columns={\"Protein ID\": \"Monomer ID\"})\n",
    "\n",
    "# make a copy of the CLNE_CLClimNE_log_data data:\n",
    "CLNE_CLClimNE_log_data_stability = CLNE_CLClimNE_log_data.copy()\n",
    "\n",
    "# remove the last three characters (like \"[c]\" for example) from the monomer IDs in the CLNE_CLClimNE_log_data data:\n",
    "CLNE_CLClimNE_log_data_stability['Monomer ID'] = CLNE_CLClimNE_log_data_stability['Monomer ID'].apply(lambda x: x[:-3] if isinstance(x, str) else x)\n",
    "\n",
    "# merge the stability data with the CLNE_CLClimNE_log_data:\n",
    "merged_df = pd.merge(CLNE_CLClimNE_log_data_stability, stability_changes, on='Monomer ID', how='left')\n",
    "\n",
    "# filter the dataframes based on the stability change:\n",
    "more_stable_df = merged_df[merged_df['Stability'] == 'More Stable']\n",
    "less_stable_df = merged_df[merged_df['Stability'] == 'Less Stable']\n",
    "leftover_df = merged_df[merged_df['Stability'].isna()]\n",
    "\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "CLNE_CLClimNE_plot_stability(fig, leftover_df, 0)\n",
    "CLNE_CLClimNE_plot_stability(fig, more_stable_df, 2)\n",
    "CLNE_CLClimNE_plot_stability(fig, less_stable_df, 1)\n",
    "\n",
    "# add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "    line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "    name=\"y=x\"));\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Protein Counts from the New WCM plotted <br>against those from the Original WCM\",\n",
    "    xaxis_title=\"log10(Orginal WCM Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600, \n",
    "    showlegend=False)\n",
    "\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "8d7b5aa5d568ed01",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Try to make the above graph not as a plotly graph, but as a matplotlib graph",
   "id": "71006aa15c5cb7cb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def CLNE_CLClimNE_plot_stability_matplotlib(ax, dataframe, stability=0):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    "    \n",
    "    if stability == 0:  # same stability\n",
    "        ax.scatter(x, y, color='grey', s=9, alpha=0.5, edgecolor='none', label=\"Same Stability\")\n",
    "    elif stability == 1:  # less stable\n",
    "        ax.scatter(x, y, color='yellowgreen', s=9, alpha=0.8, edgecolor='none', label=\"Less Stable\")\n",
    "    elif stability == 2:  # more stable\n",
    "        ax.scatter(x, y, color='lightseagreen', s=15, alpha=0.9, edgecolor='none', label=\"More Stable\")\n",
    "\n",
    "# Load the stability file from the CLClim0NE model:\n",
    "Clim_stability_data = pd.read_csv('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim0_using_biorxiv_2023_data/CLClim0NE/CLClim0NE_stability_change.csv')\n",
    "stability_changes = Clim_stability_data[[\"Protein ID\", \"Stability\"]]\n",
    "stability_changes = stability_changes.rename(columns={\"Protein ID\": \"Monomer ID\"})\n",
    "\n",
    "# Make a copy of the CLNE_CLClimNE_log_data data:\n",
    "CLNE_CLClimNE_log_data_stability = CLNE_CLClimNE_log_data.copy()\n",
    "\n",
    "# Remove the last three characters (e.g., \"[c]\") from the monomer IDs in the CLNE_CLClimNE_log_data data:\n",
    "CLNE_CLClimNE_log_data_stability['Monomer ID'] = CLNE_CLClimNE_log_data_stability['Monomer ID'].apply(lambda x: x[:-3] if isinstance(x, str) else x)\n",
    "\n",
    "# Merge the stability data with the CLNE_CLClimNE_log_data:\n",
    "merged_df = pd.merge(CLNE_CLClimNE_log_data_stability, stability_changes, on='Monomer ID', how='left')\n",
    "\n",
    "# Filter the dataframes based on the stability change:\n",
    "more_stable_df = merged_df[merged_df['Stability'] == 'More Stable']\n",
    "less_stable_df = merged_df[merged_df['Stability'] == 'Less Stable']\n",
    "leftover_df = merged_df[merged_df['Stability'].isna()]\n",
    "\n",
    "# Create a square plot and set the title\n",
    "fig, ax = plt.subplots(figsize=(6, 6), dpi=200)\n",
    "\n",
    "# Plot the scatter plots for each stability group\n",
    "CLNE_CLClimNE_plot_stability_matplotlib(ax, leftover_df, 0)\n",
    "CLNE_CLClimNE_plot_stability_matplotlib(ax, more_stable_df, 2)\n",
    "CLNE_CLClimNE_plot_stability_matplotlib(ax, less_stable_df, 1)\n",
    "\n",
    "# Add a y=x line\n",
    "ax.plot([-.25, 6], [-.25, 6], color=\"black\", linestyle=\"--\", alpha=0.2,)\n",
    "\n",
    "# Set the axis labels\n",
    "ax.set_xlabel(\"log10(WCM Protein Counts)\")\n",
    "ax.set_ylabel(\"log10(WCM w/ Gupta et al. Clim HLs Protein Counts)\")\n",
    "\n",
    "# Title and make the plot square\n",
    "ax.set_xlim(-.250, 6)\n",
    "ax.set_ylim(-.250, 6)\n",
    "plt.title(\"Updated WCM Protien Counts vs. Original WCM Protein Counts\", ha='center')\n",
    "ax.set_aspect('equal', 'box')  # Makes the plot square\n",
    "ax.legend()\n",
    "# Show plot\n",
    "plt.show()\n",
    "\n",
    "\n",
    "# save the plot as an svg file: \n",
    "fig.savefig('~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/Clim_sorting_combos/Clim0_using_biorxiv_2023_data/CLClim0NE/plots/CLNE_CLClim0NE_protein_counts_stability.svg', format='svg', dpi=1200)"
   ],
   "id": "3d423503b363a68a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Create scatter plot of CLNE counts to CLClim0NE counts, normalized by the total counts in each dataset\n",
    "# create the plot!\n",
    "def CLNE_CLClimNE_plot_with_normalization(fig, dataframe):\n",
    "    x = dataframe[\"Log10 CLNE Protein Counts\"]\n",
    "    y = dataframe[\"Log10 CLClim0NE Protein Counts\"]\n",
    "    \n",
    "    # normalize the data:\n",
    "    x = (x / sum(x)) * sum(y)\n",
    " \n",
    "    # Add scatter trace\n",
    "    fig.add_trace(go.Scatter(x=x, y=y, mode='markers', marker=dict(color='lightseagreen', size=4, opacity=.8)))\n",
    "\n",
    "    # add a y=x line\n",
    "    fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=go.scatter.Line(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"));\n",
    "    \n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "CLNE_CLClimNE_plot(fig, CLNE_CLClimNE_log_data)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"Protein Counts from the New WCM plotted <br>against those from the Original WCM\",\n",
    "    xaxis_title=\"log10(Orginal WCM Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600, \n",
    "    showlegend=False)\n",
    "\n",
    "# Show figure\n",
    "fig.show()"
   ],
   "id": "73d56d4e83d5ac40",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Plot B: plot of CLClime0NE model (with the protease substrates highlighted) plotted against Schmidt data",
   "id": "c3f16e91ab63c3a4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# TODO: should this be normalized?\n",
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "# Create scatter plot of Schmidt Validation Protein Counts vs CLNE and CLClimNE Counts\n",
    "def add_scatter_and_trendline(fig, df, name, is_protease_substrate=False):\n",
    "    x = df[\"Log10 Validation Schmidt Counts\"]\n",
    "    y = df[\"Log10 Simulated Schmidt Counts\"]\n",
    "    hovertext = df[\"gene symbol\"]\n",
    "\t\n",
    "    # Properly indented if statement\n",
    "    if is_protease_substrate == True:\n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', \n",
    "                                  marker_size=7, marker=dict(color=\"orange\", symbol='star'), \n",
    "                                  name=name))\n",
    "    else: \n",
    "        fig.add_trace(go.Scatter(x=x, y=y, hovertext=hovertext, mode='markers', marker_size=3,\n",
    "                                 marker=dict(color=\"lightseagreen\", opacity=.5),  name=name))\n",
    "\t\t\n",
    "        \n",
    "# Create figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add scatter plots and trendlines for each dataset\n",
    "add_scatter_and_trendline(fig, CLClim0NE_VS_log_data_remaining, (current_sequence + ' vs Schmidt'))\n",
    "add_scatter_and_trendline(fig, CLClim0NE_protease_substrates, (current_sequence + '(protease substrates) vs Schmidt'), True)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=f\"New WCM (including Clim half-lives) Protein Counts <br>plotted against Validation Protein Counts\",\n",
    "    xaxis_title=\"log10(Validation Protein Counts)\",\n",
    "    yaxis_title=f\"log10(New WCM Protein Counts)\",\n",
    "    autosize=False,\n",
    "    width=600,\n",
    "    height=600, \n",
    "    showlegend=False)\n",
    "\n",
    "\n",
    "# Add a y=x line\n",
    "fig.add_trace(go.Scatter(x=[0, 6], y=[0, 6], mode=\"lines\",\n",
    "        line=dict(color=\"black\", dash=\"dash\"), opacity=0.2,\n",
    "        name=\"y=x\"))\n",
    "\n",
    "# Show figure\n",
    "fig.show()\n"
   ],
   "id": "a5a68400234a4563",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "683db366d4dc42ab",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# working with Supplementary data #6: \"TableS6 - Absolute protein levels >>\"\n",
    "\n",
    "Importing this data to see how it holds up against the protein count data we got from our simulation"
   ],
   "id": "df678a5fd4b09660"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# load in the absolute protein concentration data from Gupta, et. al., 2024: \n",
    "ST6_PC_data = pd.read_excel(\"~/wcEcoli/models/ecoli/analysis/local_notebooks/C_limited_PDR_analyses_with_published_paper_data/supplementary_data/Gupta_et_al_2024_ST6_absolute_protein_levels.xlsx\")\n",
    "\n",
    "# take the log of the of the concentration: \n",
    "ST6_PC_data[\"log10 concentration\"] = np.log10(ST6_PC_data[\"Concentration (M)\"])\n",
    "\n",
    "ST6_PC_data\n",
    "# going to have to work on this later, because the concentration units are not the same and it might be for the N-limited connditions. \n",
    "sum(ST6_PC_data[\"Concentration (M)\"])\n",
    "\n",
    "# might want to: \n",
    "# - 1: compare ST6 against my total protein counts (not log version) and figure out which proteins I have match reads for\n",
    "# - 2: then, normalize one to the other by dividing each protein's counts by the total counts, and then multiplying by the sum of the total counts for the other dataset "
   ],
   "id": "85577a74e0c3be10",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
